src/validation/validation.cpp

```diff
@@ -4326,61 +4327,93 @@ bool StartPartialHeaders(int64_t time, const std::function<void(const CBlockInde
+    }
 
-    if (    IsPartialSyncActive()
-         && youngestBefore && (youngestBefore->nHeight <= 576 || youngestBefore->nHeight - partialChain.HeightOffset() > 576))
+    if (IsPartialSyncActive() && youngestBefore && (youngestBefore->nHeight <= 576 || youngestBefore->nHeight - partialChain.HeightOffset() > 576))
     {
         LogPrintf("Partial sync continues.\n");
     }
-    else {
+    else
+    {
         if (IsPartialSyncActive()) // IsPartialSyncActive() => above checks for time and/or 576 window failed
         {
             LogPrintf("Partial sync in progress but starting point is too young for requested start. Sync reset.\n");
             ResetPartialSync();
         }
 
-        // Search checkpoint with at least a 576 block height difference to the checkpoint that is the youngest before
-        // the desired starting time. This way when we reach the latter checkpoint there will be enough data to do context checks
-        // the largest windows for context checking is the DELTA algorithm, which needs 576 blocks.
-        CheckPointEntry entry;
-        int youngestHeight = Checkpoints::LastCheckpointAt(time, entry);
-        int olderHeight = youngestHeight;
-        while (olderHeight > 0 && youngestHeight - olderHeight <= 576)
+
+        // Determine the first checkpoint that comes before wallet birth date
+        uint64_t nWalletBirthBlockHard = Checkpoints::LastCheckpointBeforeTime(time);
+
+        // Now determine the first checkpoint of actual interest using block filters
+        uint64_t nWalletBirthBlockSoft = Checkpoints::LastCheckPointHeight();
+        if (pactiveWallet)
         {
-            olderHeight = Checkpoints::LastCheckpointAt(entry.nTime - 1, entry);
+            LOCK2(cs_main, pactiveWallet?&pactiveWallet->cs_wallet:NULL);
+
+            GCSFilter::ElementSet elementSet;
+            for (const auto& [accountUUID, forAccount] : pactiveWallet->mapAccounts)
+            {
+                (unused) accountUUID;
+                std::set<CKeyID> setAddresses;
+                forAccount->GetKeys(setAddresses);
+                for (const auto& key : setAddresses)
+                {
+                    //fixme: (2.1) Alter the blockfilter to use just the keys instead of the full script
+                    //This will be more efficient in terms of space and time and simplify code like this
+                    CScript searchScript = CScript() << OP_DUP << OP_HASH160 << ToByteVector(key) << OP_EQUALVERIFY << OP_CHECKSIG;
+                    std::vector<unsigned char> searchData(searchScript.begin(), searchScript.end());
+                    elementSet.insert(searchData);
+                }
+            }
+            std::vector<std::tuple<uint64_t, uint64_t>> cpRanges;
+            getBlockFilterBirthAndRanges(nWalletBirthBlockHard, nWalletBirthBlockSoft, elementSet, cpRanges);
+            std::swap(cpRanges, partialChain.cpRanges);
```

Comment:
We need to call this again as new addresses are generated during sync as well (if e.g. a transaction is found)